README.md
# mysql probe

A golang application that checks a mysql database, writing results to flat files and/or runs an HTTP server for haproxy health checks based on those checks

# Requirements

* MySQL >= 5.1.12
* Access to a MySQL user with REPLICATION CLIENT permissions (or ability to create one)

## Installation

1. Install golang 1.3+
2. `go get github.com/haikulearning/mysql_probe`

### MySQL configuration

Create a MySQL user with the following permissions:
* REPLICATION CLIENT

For example:

````sql
CREATE USER 'mysql_probe'@'127.0.0.1' IDENTIFIED BY 'As92rj05UvKK';
GRANT REPLICATION CLIENT ON *.* TO 'mysql_probe'@'127.0.0.1';
````

## Running

````
$ mysql_probe --help # will provide available commands & flags.
$ mysql_probe start  # will both test mysql and run the HTTP status server.
````

### Checks

For now, a default set of checks are hard-coded to run:
* "Connection Count Less Than X" with the following values:
  * 1, 10, 50, 100, 150, 200, 250, 300, 600, 1200, 2400
* "Replication Delay Less Than X" with the following values (in seconds):
  * 0, 10, 30, 60, 120, 300, 600, 1200, 2400

### haproxy Configuration Example
TOOD: Add an example haproxy config using these checks to build a resilient failover configuration.

### xinetd Configuration Example
TODO: Add an example of using xinetd to return these http txt files as http requests.

## Development

````bash
$ get go -u github.com/haikulearning/mysql_probe
$ cd $GOPATH/src/github.com/haikulearning/mysql_probe
$ make
````

### Cross-compiling

If you need to deploy `mysql_probe` to another platform (i.e. build on OS X, deploy to Linux), Go makes that fairly straightforward!

````bash
$ cd $GOPATH/src/github.com/haikulearning/mysql_probe
$ make all
# or target a specific OS
$ make mac
$ make linux
````

If the above commands aren't working for you, be you've locally installed appropriate Go compilers for the target platforms (here's a Stack Overflow question [describing how to install them on OS X](http://stackoverflow.com/questions/12168873/cross-compile-go-on-osx)).

## Testing
A dockerfile is available in docker/test for spinning up an instance of this app.

## Roadmap
1. Build checks with completely hardcoded configuration to check localhost.
2. Build configuration parser so this app does not need to be recompiled every time config changes.
3. Build rollup checks into configuration.

### Phase 1 hardcoded config
Check config.go for hardcoded values, host, username, password.
Run replcation_delay check and output to /var/run/mysql_probe/replcation_delay.http.txt
Run connection_count check and output to /var/run/mysql_probe/connection_count.http.txt

## Copyright

Copyright 2014 Derek Volker

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

Because this project has not reached 1.0.0 yet anything may change at any time. The public API should not be considered stable. Still, all breaking changes to the public API will be documented in this file.

## [0.3.0] - 2015-09-04
### Breaking Changes
* Renamed `--jsonlog` to `--testlog`
* Changed default value of `--testlog` to `tmp/test.log`, because it's *so*, so verbose for STDOUT!

### Added
* Added a `--serverlog` option so that minimal info is written to STDOUT
* An initial, basic version of `make release`


## [0.2.0] - 2015-08-27
### Breaking Changes
* `connection_count_lte_*` report file will no longer be created. Use `threads_connected_count_lte_*` report files instead (they're functionally equivalent).

### Added
* This Change Log!
* Initial support for cross-compiling via make

### Changed
* Counting the # of connections has been refactored via [Pull #3](https://github.com/haikulearning/mysql_probe/pull/3) so that fewer permissions are required by the `mysql_probe`'s MySQL user.
* Major improvements to the Makefile inspired by the [akitu/github-release project's](https://github.com/aktau/github-release) Makefile

## [0.1.0] - 2015-08-27
### Breaking Changes
* Use `mysql_probe start` instead of just `mysql_probe`

### Added
* Use `mysql_probe test` or `mysql_probe serve` to run only the test or status server functionality, respectively.
* Status server supports `--reports` and `--server_port` flags.
* A very basic Makefile

## [0.0.3] - 2015-08-20
### Added
* A basic status HTTP server
* Tests time out if MySQL is slow to respond

### Changed
* Moved the testing logic of `mysql_probe` into a background Goroutine.

## [0.0.2] - 2014-12-18
### Added
* JSON logs
* Reports directory

## [0.0.1] - 2014-12-18
### Added
* "Everything!" This was the initial release.


[unreleased]: https://github.com/haikulearning/mysql_probe/compare/v0.3.0...HEAD
[0.3.0]: https://github.com/haikulearning/mysql_probe/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/haikulearning/mysql_probe/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/haikulearning/mysql_probe/compare/v0.0.3...v0.1.0
[0.0.3]: https://github.com/haikulearning/mysql_probe/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/haikulearning/mysql_probe/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/haikulearning/mysql_probe/compare/0c36901b85f8e...v0.0.1

main.go
/*
   Copyright 2014 Derek Volker

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

package main

import (
	"github.com/codegangsta/cli"
	"github.com/haikulearning/mysql_probe/mysqltest"
	"github.com/haikulearning/mysql_probe/statusserver"
	"os"
  "log"
  "sync"
)

// The current version of the app
const VERSION string = "0.3.0"

func test_mysql(c *cli.Context) {
  log.Println("Testing mysql server")
  // Run our mysql tests
  mysqltest.RunMysqlTest("connection", c.String("host"), c.Int("port"), c.String("user"), c.String("pass"), c.Int("interval"), c.Int("timeout"), c.String("reports"), c.String("testlog"))
}

func serve_status(c *cli.Context) {
  log.Println("Running status server")
  // Run our status server
  statusserver.StartStatuServer(c.String("reports"), c.Int("server_port"), c.String("serverlog"))
}

func test_and_serve(c *cli.Context) {
  var wg sync.WaitGroup

  wg.Add(1)
  go func() {
    // Decrement the counter when the goroutine completes.
    defer wg.Done()
    test_mysql(c)
  }()

  wg.Add(1)
  go func() {
    // Decrement the counter when the goroutine completes.
    defer wg.Done()
    serve_status(c)
  }()

  wg.Wait()
}

func main() {
	app := cli.NewApp()
	app.Name = "mysql_probe"
	app.Usage = "test mysql health, write the results to disk, and serve them via http"
	app.Version = VERSION

  test_mysql_flags := []cli.Flag{
		cli.StringFlag{
			Name:   "host",
			Value:  "127.0.0.1",
			Usage:  "(test) mysql host to connect to",
			EnvVar: "MYSQL_PROBE_HOST",
		},
		cli.IntFlag{
			Name:   "port, p",
			Value:  3306,
			Usage:  "(test) mysql port to connect to",
			EnvVar: "MYSQL_PROBE_PORT",
		},
		cli.StringFlag{
			Name:   "user, u",
			Value:  "root",
			Usage:  "(test) mysql username to connect with",
			EnvVar: "MYSQL_PROBE_USER",
		},
		cli.StringFlag{
			Name:   "pass",
			Value:  "test",
			Usage:  "(test) mysql password to connect with",
			EnvVar: "MYSQL_PROBE_PASS",
		},
		cli.IntFlag{
			Name:   "timeout, t",
			Value:  2000,
			Usage:  "(test) time in milliseconds to wait for a mysql connection",
			EnvVar: "MYSQL_PROBE_TIMEOUT",
		},
		cli.IntFlag{
			Name:   "interval, i",
			Value:  250,
			Usage:  "(test) interval in milliseconds to run the checks, set to 0 to only run the tests once",
			EnvVar: "MYSQL_PROBE_INTERVAL",
		},
		cli.StringFlag{
			Name:   "testlog",
			Value:  "tmp/test.log",
			Usage:  "(test) file to write test results log output in json",
			EnvVar: "MYSQL_PROBE_TESTLOG",
		},
  }
  both_flags := []cli.Flag{
		cli.StringFlag{
			Name:   "reports",
			Value:  "tmp",
			Usage:  "(all) directory for test results up/down status files",
			EnvVar: "MYSQL_PROBE_REPORTS",
		},
  }
  server_status_flags := []cli.Flag{
		cli.IntFlag{
			Name:   "server_port, s",
			Value:  3001,
			Usage:  "(serve) port number where the status server accepts requests",
			EnvVar: "MYSQL_PROBE_SERVERPORT",
		},
		cli.StringFlag{
			Name:   "serverlog",
			Value:  "tmp/server.log",
			Usage:  "(test) file to write server access data in json",
			EnvVar: "MYSQL_PROBE_SERVERLOG",
		},
  }

  all_flags := []cli.Flag{}
  all_flags = append(all_flags, test_mysql_flags...)
  all_flags = append(all_flags, both_flags...)
  all_flags = append(all_flags, server_status_flags...)

  app.Commands = []cli.Command{
    {
      Name: "start",
      Usage: "both test mysql & run a status server of the results",
      Action: func(c *cli.Context) {
        test_and_serve(c)
      },
      Flags: all_flags,
    },
    {
      Name: "test",
      Usage: "test a mysql server's status",
      Action: func(c *cli.Context) {
        test_mysql(c)
      },
      Flags: append(test_mysql_flags, both_flags...),
    },
    {
      Name: "serve",
      Usage: "run server of the test's output files",
      Action: func(c *cli.Context) {
        serve_status(c)
      },
      Flags: append(server_status_flags, both_flags...),
    },
  }
	app.EnableBashCompletion = true
	app.Run(os.Args)

	// TODO: write output files as failures on termination
}


jsonlog/jsonlog.go
package jsonlog

import (
  "fmt"
	"os"
	"encoding/json"
	"time"
)

type JsonLog struct {
	jsonlog    *os.File
}

func Init(log_path string) *JsonLog {
	open_file, err := os.OpenFile(log_path, os.O_WRONLY|os.O_CREATE|os.O_APPEND, 0644)
	if err != nil {
		panic(fmt.Sprintf("Couldn't open log_path \"%s\" for writing: %s", log_path, err.Error()))
	}

	l := JsonLog{jsonlog: open_file}

	return &l
}

func (l *JsonLog) LogFile() *os.File {
	return l.jsonlog
}

func (l *JsonLog) Log(msg string, other_data interface{}) {
	data, err := json.Marshal(other_data)
	if err != nil {
		panic(err.Error())
	}

	l.jsonlog.WriteString(
		fmt.Sprintf("{\"@timestamp\":\"%s\",\"type\":\"mysql_probe\",\"message\":\"%s\",\"data\":",
			time.Now().Format(time.RFC3339), msg))
	l.jsonlog.Write(data)
	l.jsonlog.WriteString("}\n")
}


statusserver/server.go
package statusserver

import (
  "strconv"
  "fmt"
  "log"
  "net/http"
  "os"
  "regexp"
  "github.com/haikulearning/mysql_probe/mysqltest"
	"github.com/haikulearning/mysql_probe/jsonlog"
)

var required_up_checks = []string{"connect", "threads_connected_count_lte_2400"}

type StatuServer struct {
	reportdir     string
	port          int
	jsonlog       *jsonlog.JsonLog
}

func StartStatuServer(reportdir string, port int, log_path string) *StatuServer {

	jsonlog := jsonlog.Init(log_path)
	defer jsonlog.LogFile().Close()

	s := StatuServer{reportdir: reportdir, port: port, jsonlog: jsonlog}

  s.Start()

	return &s
}

// Start up the status server
func (s *StatuServer) Start() {
  http.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
    s.handler(w, r)
  })
  //fs := http.FileServer(http.Dir("tmp"))
  //http.Handle("/", fs)

  log.Println("Listening to port " + strconv.Itoa(s.port))
  http.ListenAndServe(":" + strconv.Itoa(s.port), nil)
}

func check(e error) {
    if e != nil {
        panic(e)
    }
}

func (s *StatuServer) handler(w http.ResponseWriter, r *http.Request) {
  is_up := true

  for _,testname := range required_up_checks {
    if is_up {
      is_up = s.testResultIsUp(testname, r)
    }
  }

  if is_up {
    w.WriteHeader(http.StatusOK)
    fmt.Fprintf(w, "up\n")
  } else {
    w.WriteHeader(http.StatusServiceUnavailable)
    fmt.Fprintf(w, "down\n")
  }
}

func (s *StatuServer) testResultIsUp(testname string, r *http.Request) bool {
  testpath := mysqltest.TestResultPath(s.reportdir, testname)
  match := false

  defer func() {
    if r := recover(); r != nil {
      log.Println("down via failure checking " + testpath + " (skipping all subsequent checks)")
    }
  }()

  f, err := os.Open(testpath)
  check(err)

  b1 := make([]byte, 10)
  _, err = f.Read(b1)
  check(err)

  match, err = regexp.MatchString("up", string(b1))
  check(err)

  if match {
    s.Log("up via " + testpath, r)
  } else {
    s.Log("down via " + testpath + " (skipping all subsequent checks)", r)
  }

  return match
}

func (s *StatuServer) Log(msg string, r *http.Request) {
	json_msg := map[string]interface{}{
		"Method": 		r.Method,
		"RequestURI": r.RequestURI,
		"RemoteAddr": r.RemoteAddr,
	}

	s.jsonlog.Log(msg, json_msg)
}


mysqltest/mysqlTestResultWriter.go
package mysqltest

import (
  "os"
	"fmt"
	"path/filepath"
)

type MysqlTestResultWriter struct {
  test    *MysqlTest
  result  *MysqlTestResult
}

func TestResultPath(reportdir string, testname string) string {
  return fmt.Sprintf("%s.agent.txt", filepath.Join(reportdir, "/", testname))
}

// writes corresponding files for each test result entry
func (tw *MysqlTestResultWriter) WriteResult() {
  for k,v := range tw.result.Entries {
    tw.WriteTextResult(k,v)
  }
}

// write an individual test result into to a text file
func (tw *MysqlTestResultWriter) WriteTextResult(testname string, status string) {
	path := TestResultPath(tw.test.reportdir, testname)
	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
	if err != nil {
		//log.Fatal(err)
		desc := fmt.Sprintf("Couldn't open result file \"%s\": ", err.Error())
		tw.test.Log(desc)
		os.Stderr.WriteString(desc)
		return
	}
	defer file.Close()
	tw.test.Log(fmt.Sprintf("Test: %s result: %s", testname, status))
	file.WriteString(status)
}

//func (t *MysqlTest) WriteHttpResult(testname string, passed bool, description string) {
//	status := "503 Service Unavailable"
//	if passed {
//		status = "200 OK"
//	}
//	now := time.Now().Format(time.RFC1123Z)
//	response := fmt.Sprintf("HTTP/1.1 %s\r\nDate: %s\r\nContent-Type: text/plain\r\n\r\n%s\r\n", status, now, description)
//
//	path := fmt.Sprintf("%s.http.txt", filepath.Join(t.reportdir, "/", testname))
//	file, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0644)
//	if err != nil {
//		//log.Fatal(err)
//		desc := fmt.Sprintf("Couldn't open result file \"%s\": ", err.Error())
//		tw.test.JsonLog(desc)
//		os.Stderr.WriteString(desc)
//		return
//	}
//	defer file.Close()
//	tw.test.JsonLog(fmt.Sprintf("Test: %s result: %v: %s", testname, passed, description))
//	file.WriteString(response)
//}

mysqltest/mysqlTest.go
/*
   Copyright 2014 Derek Volker

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/

package mysqltest

import (
	"database/sql"
	"fmt"
	_ "github.com/go-sql-driver/mysql"
	"github.com/haikulearning/mysql_probe/jsonlog"
	"errors"
	"time"
)

var counts_to_check = []int64{1, 10, 50, 100, 150, 200, 250, 300, 600, 1200, 2400}
var seconds_to_check = []int64{0, 10, 30, 60, 120, 300, 600, 1200, 2400}

type MysqlTest struct {
	Name          string
	reportdir     string
	host          string
	user          string
	port          int
	pass          string
	interval      int
	timeout       int
	db            *sql.DB
	jsonlog       *jsonlog.JsonLog
	iteration     uint64
	maxjobs       uint
}


func RunMysqlTest(name string, host string, port int, user string, pass string, interval int, timeout int, reportdir string, log_path string) *MysqlTest {

	jsonlog := jsonlog.Init(log_path)
	defer jsonlog.LogFile().Close()


	m := MysqlTest{Name: name, host: host, port: port, user: user, pass: pass, interval: interval, timeout: timeout, reportdir: reportdir, jsonlog: jsonlog, iteration: 1}

	m.Run()

  return &m
}

func (t *MysqlTest) Run() {
	if t.interval < 0 {
		panic("interval must be a positive integer, or 0 to run the tests only once.")
	}
	t.RunOnceWithTimeout() // Run first test instantly.
	if t.interval > 0 {
		// run checks on intervals
		for _ = range time.Tick(time.Duration(t.interval) * time.Millisecond) {
			t.iteration++ // when this overflows it will become 0 with no problems http://play.golang.org/p/fbjwHKcIaU
			t.RunOnceWithTimeout()
		}
	}
}

func (t *MysqlTest) GetWeight(val int64, max int64) string {
	if val == 0 {
		return "100%"
	}
	return fmt.Sprintf("%d%%", 100-(100*(val/max)))
}

func (t *MysqlTest) RunOnceWithTimeout() {
  timeout_ch := make(chan *MysqlTestResult, 1)
  ch := make(chan *MysqlTestResult, 1)

  // after 1 second, fill a blank response and
  // send that indicates everything timed out
  go func() {
    time.Sleep(time.Duration(t.timeout) * time.Millisecond)
    // fill timeout statuses up
    res := MysqlTestResult{}
    res.AddTextResult("connect", "down # timeout")
    for _,c := range counts_to_check {
      res.AddTextResult(fmt.Sprintf("threads_connected_count_lte_%d", c), "down # timeout")
      res.AddTextResult(fmt.Sprintf("threads_running_count_lte_%d", c), "down # timeout")
    }
    for _,c := range seconds_to_check {
       res.AddTextResult(fmt.Sprintf("replication_delay_lte_%d", c), "down # timeout")
    }
    timeout_ch <- &res
  }()

  // run the actual check with no delay
  // hopefully it wins.
  go func() {
    res := t.RunOnce()
    ch <- res
  }()

  var res * MysqlTestResult
  // grab the first thing that comes back
  select {
  case res = <-ch:
  case res = <-timeout_ch:
  }

  writer := MysqlTestResultWriter{test: t, result: res}
  writer.WriteResult()
}

func (t *MysqlTest) RunOnce() *MysqlTestResult {
        res := MysqlTestResult{}
	start := time.Now()
	err := t.Connect()
	dur := time.Since(start)
	if err != nil {
		res.AddTextResult("connect", fmt.Sprintf("down # %s", err.Error()))
	} else {
		res.AddTextResult("connect", fmt.Sprintf("up 100%% # connect took %s", dur.String()))
	}

	start = time.Now()
	threads_connected, threads_running, description, err := t.CountThreadInfo()
	if err != nil {
		description = fmt.Sprintf("%s: %s", description, err.Error())
		for _, connections := range counts_to_check {
			res.AddTextResult(fmt.Sprintf("threads_connected_count_lte_%d", connections), fmt.Sprintf("down # Threads Connected count test (%d connected <= %d)? : %s", threads_connected, connections, description))
			res.AddTextResult(fmt.Sprintf("threads_running_count_lte_%d", connections), fmt.Sprintf("down # Threads Running count test (%d running <= %d)? : %s", threads_running, connections, description))
		}
	} else {
		for _, connections := range counts_to_check {
			var status string

			if threads_connected <= connections {
				status = fmt.Sprintf("up %s", t.GetWeight(threads_connected, connections))
			} else {
				status = "down"
			}
			res.AddTextResult(fmt.Sprintf("threads_connected_count_lte_%d", connections), fmt.Sprintf("%s # Threads Connected count test (%d connected <= %d)?", status, threads_connected, connections))

			if threads_running <= connections {
				status = fmt.Sprintf("up %s", t.GetWeight(threads_running, connections))
			} else {
				status = "down"
			}
			res.AddTextResult(fmt.Sprintf("threads_running_count_lte_%d", connections), fmt.Sprintf("%s # Threads Connected count test (%d connected <= %d)?", status, threads_running, connections))
		}
	}

	delay, description, err := t.CheckReplication()
	if err != nil {
		description = fmt.Sprintf("%s: %s", description, err.Error())
		for _, seconds := range seconds_to_check {
			res.AddTextResult(fmt.Sprintf("replication_delay_lte_%d", seconds), fmt.Sprintf("down # Replication delay test (%d delay <= %d seconds)? : %s", delay, seconds, description))
		}
	} else {
		for _, seconds := range seconds_to_check {
			var status string
			if delay <= seconds {
				status = fmt.Sprintf("up %s", t.GetWeight(delay, seconds))
			} else {
				status = "down"
			}
			res.AddTextResult(fmt.Sprintf("replication_delay_lte_%d", seconds), fmt.Sprintf("%s # Replication delay test (%d delay <= %d seconds)? : %s", status, delay, seconds, description))
		}
	}
	defer t.Disconnect()

  return &res
}

func (t *MysqlTest) Disconnect() {
	//fmt.Println("Disconnecting")
	if t.db != nil {
		t.db.Close()
	}
}

// NOTE: only works if one master
func (t *MysqlTest) CheckReplication() (int64, string, error) {
	var description string
	if t.db == nil {
		description = "DB connection is invalid"
		return 0, description, errors.New(description)
	}
	rows, err := t.db.Query("SHOW SLAVE STATUS") // queryable from PERFORMANCE_SCHEMA at mysql 5.7.2: http://bugs.mysql.com/bug.php?id=35994
	if err != nil {
		description = "Query 'SHOW SLAVE STATUS' failed"
		t.Log(fmt.Sprintf("%s: %s", description, err.Error()))
		t.db = nil
		return 0, description, err
	}

	// Since our mysql is too old to use PERFORMANCE_SCHEMA and select only the columns we want,
	// we need to find the column count and use RawBytes to be a placeholder for all the columns we don't care about.
	// http://go-database-sql.org/varcols.html
	cols, err := rows.Columns() // Remember to check err afterwards
	if err != nil {
		description = "Couldn't retrieve column information for 'SHOW SLAVE STATUS' statement"
		t.Log(fmt.Sprintf("%s: %s", description, err.Error()))
		t.db = nil
		return 0, description, err
	}
	vals := make([]interface{}, len(cols))
	var slave_io_running string
	var seconds_behind_master int64
	seconds_behind_master = -1
	for i, _ := range cols {
		switch cols[i] {
		case "Slave_IO_Running":
			vals[i] = &slave_io_running
		case "Seconds_Behind_Master":
			vals[i] = &seconds_behind_master
		default:
			vals[i] = new(sql.RawBytes)
		}
	}
	err = nil
	for rows.Next() {
		err = rows.Scan(vals...)
		// Now you can check each element of vals for nil-ness,
		// and you can use type introspection and type assertions
		// to fetch the column into a typed variable.
	}
	if slave_io_running == "" || seconds_behind_master == -1 {
		description = fmt.Sprintf("Slave status and/or seconds_behind_master could not be determined. slave_io_running: %s, seconds_behind_master: %d", slave_io_running, seconds_behind_master)
		err = errors.New(description)
	} else if slave_io_running == "No" {
		description = "Slave is not running"
		err = errors.New(description)
	} else {
		description = fmt.Sprintf("Slave running: %s Seconds behind: %d", slave_io_running, seconds_behind_master)
	}

	t.Log(description)
	return seconds_behind_master, description, err
}

func (t *MysqlTest) CountThreadInfo() (int64, int64, string, error) {
  var description string

	if t.db == nil {
		return -2, -2, "No DB connection available", nil
	}

  // INFORMATION_SCHEMA.SESSION_STATUS introduced in MySQL 5.1.12
  // http://dev.mysql.com/doc/refman/5.1/en/status-table.html
	rows, err := t.db.Query("SELECT VARIABLE_NAME, VARIABLE_VALUE FROM INFORMATION_SCHEMA.SESSION_STATUS WHERE VARIABLE_NAME IN ('THREADS_CONNECTED', 'THREADS_RUNNING')")
	if err != nil {
		description = "Query 'SELECT ... FROM INFORMATION_SCHEMA.SESSION_STATUS ...' failed"
		t.Log(fmt.Sprintf("%s: %s", description, err.Error()))
		t.db = nil
		return -1, -1, description, err
	}

	var threads_connected int64
	var threads_running int64
	var variable_name string
	var variable_val int64
	for rows.Next() {
		err = rows.Scan(&variable_name, &variable_val)
		switch variable_name {
		case "THREADS_CONNECTED":
			threads_connected = variable_val
		case "THREADS_RUNNING":
			threads_running = variable_val
		}
	}

	threads_connected = threads_connected - 1 // Deduct this connection from teh count.

	description = fmt.Sprintf("Threads Connected: %d; Threads Running: %d", threads_connected, threads_running)
	t.Log(description)
	return threads_connected, threads_running, description, nil
}

func (t *MysqlTest) Connect() error {
	t.Log(fmt.Sprintf("Connecting to %s@%s:%d", t.user, t.host, t.port))

	// Create dsn like such https://github.com/Go-SQL-Driver/MySQL/#dsn-data-source-name
	// [username[:password]@][protocol[(address)]]/dbname[?param1=value1&...&paramN=valueN]
	// username:password@protocol(address)/dbname?param=value
	dsn := fmt.Sprintf("%s:%s@tcp(%s:%d)/?timeout=%dms&allowOldPasswords=1", t.user, t.pass, t.host, t.port, t.timeout)
	//fmt.Println("DSN: ", dsn)

	db, err := sql.Open("mysql", dsn)
	if err != nil {
		t.Log(fmt.Sprintf("Couldn't open mysql connection: %s", err.Error()))
		t.db = nil
		return err
	}

	// Open doesn't open a connection. Validate DSN data:
	err = db.Ping()
	if err != nil {
		t.Log(fmt.Sprintf("Couldn't connect to mysql server: %s", err.Error()))
		t.db = nil
		return err
	}

	//fmt.Println("SUCCESS")
	t.db = db
	db.Query("SET SESSION wait_timeout=1")
	return nil
}

func (t *MysqlTest) Log(msg string) {
	json_msg := map[string]interface{}{
		"host": t.host,
		"iteration": t.iteration,
	}

	t.jsonlog.Log(msg, json_msg)
}
